crates/api/src/trampoline/func.rs

```diff
@@ -87,24 +88,38 @@ unsafe extern "C" fn stub_fn(
     };
 
     let mut returns = vec![Val::null(); returns_len];
-    let func = &instance
+    let state = &instance
         .host_state()
         .downcast_mut::<TrampolineState>()
-        .expect("state")
-        .func;
-
-    match func.call(&args, &mut returns) {
-        Ok(()) => {
+        .expect("state");
+
+    // Be sure to `catch_unwind` here in case our callable panics, and if so we
+    // need to manually carry the panic across the JIT frames back to the
+    // original call-site of wasm code. Note that eventually we may be able to
+    // JIT code such that native unwinding can unwind through those frames, but
+    // for now it's just easier to always catch panics.
+    let result = panic::catch_unwind(AssertUnwindSafe(|| state.func.call(&args, &mut returns)));
+    match result {
+        // On success we write out all the results into the return pointer that
+        // we were given.
+        Ok(Ok(())) => {
             for (i, r#return) in returns.iter_mut().enumerate() {
                 // TODO check signature.returns[i].value_type ?
                 r#return.write_value_to(values_vec.add(i));
             }
-            0
-        }
-        Err(trap) => {
-            record_api_trap(trap);
-            1
         }
+
+        // If a trap was raised (an error returned from the imported function)
+        // then we smuggle the trap through `Box<dyn Error>` through to the
+        // call-site, which gets unwrapped in `Trap::from_jit` later on as we
+        // convert from the internal `Trap` type to our own `Trap` type in this
+        // crate.
+        Ok(Err(trap)) => wasmtime_runtime::raise_user_trap(Box::new(trap)),
+
+        // And finally if the imported function panicked, then we trigger the
+        // form of unwinding that's safe to jump over wasm code on all
+        // platforms.
+        Err(panic) => wasmtime_runtime::resume_panic(panic),
```

Comment:
I wonder if this is safe? This call ultimately leads to a longjmp so we would just jettison from this frame while there are some values that are `Drop`.

Reply:
Should we not simply translate a panic to a `Trap` representation here and return 1 so we can let the calling trampoline code trap?

Do we really need to resume the panic on the other side of the Wasm frames?

When Wasm supports exception handling in the future, do we want to make panicking imported host functions uncatchable?

Personally, I feel that a faulting host functions should be treated like they trapped.

Reply:
One more thing to consider regarding treating panicking host functions as traps: if I have a host running multiple Wasm "microprocesses" simultaneously, I sure would like the runtime to abort the trapping "microprocess" and not the host process just because one of those Wasm programs hit a bug in the host.

Reply:
Ok, last comment regarding traps: there might be foreign frames we don't want unwound via a panic.

For example, a .NET host might implement a host function that calls back into wasm, which calls into a Rust implemented host import (Wasmtime's WASI implementation, for example).

If that top Rust frame panics and we unwind, we're unwinding *through* the foreign .NET frames.  That's no bueno.

The bottom line is we can't guarantee the stack is always a mix of Rust and Wasm frames.

Reply:
Excellent points! I think these are two separate questions, though, so I'll try to answer both separately:

> I wonder if this is safe? This call ultimately leads to a longjmp so we would just jettison from this frame while there are some values that are Drop.

Oh good eyes, I forgot about that! I do believe that this is safe in the memory safety sense, but it's wrong in that we don't want to leak any data. I've refactored a bit to avoid the leaks by ensuring that when we longjmp there's no variables on the stack. I believe from a safety perspective this should be all A-OK. I'll leave some comments about the unsafety on the raise functions though.

> do we really need to resume the panic on the other side of the Wasm frames

This I do think is an imported API design consideration, and is very much so worth exploring! I think the main "prior art" that we could possibly have is JS, but there aren't necessarily great analogs. If an imported function throws an exception in JS it simply throws the exception past all the wasm into the normal JS. If wasm traps it turns into a JS exception as well.

In Rust though we sort of have two ways of communicating errors (rather than the one that JS has). One is with `Result` and one is with panics. Using `Result` is intended for "normal expected errors" whiel `panic` typically means "a programmer error" (aka a bug). This is similar (ish) to return codes in C vs `assert`. (note that I say all this to basically say that the prior art of JS may not guide us much because there's only one choice there, while in Rust we have two).

I personally think though that propagation of panics is probably what we want here. Ideally we wouldn't even need to catch the panic at all, we'd simply have the system unwinder fly past all our JIT frames. I don't believe that works right now though (not without extra effort), and I wanted to get things working first. In a world though where we emitted correct exception handling tables and such then if we were to translate panics to traps we would have to unconditionally catch the panic no matter what.

There's also the complication of `panic=unwind` and `panic=abort` in Rust as well. This is sort of where panics in Rust, seen as programmer errors, can be determined at compile time to default destroy the current thread or the entire process. (and the default in rustc is the current thread). Functions like `catch_unwind`, however, have no effect in `panic=abort` mode, which means that `wasmtime` wouldn't be able to provide the guarantee that we could translate panics to traps as well.

I also feel that `Trap` vs `panic!` is a good distinction to keep. I sort of would expect a runtime bug to by default tear everything down quickly whereas a `Trap` would be properly handled to tear down a microprocess.

Anyway that's sort of my thinking, but curious how you feel given all that info @peterhuene?

Reply:
Oh I finished typing before your latest comment @peterhuene, but I think that's also a really good point. In general it is definitely, yeah, undefined for Rust panics to fly through anything other than Rust frames. I would consider this the responsibility of the embedder, however, which is embedding Rust code to handle that. Typically though I would also expect that projects would compile with `panic=abort` so a panic in Rust would abort the process anyway.

Reply:
I disagree that any panic originating from the runtime or a host-defined function should abort the entire process, assuming it's possible to avoid.  I think there's a problem in our implementation if we can't provide for simultaneous "microprocess" execution and properly isolate those processes from one another.

Think of it purely from an operating systems perspective: if my code is running in one isolated process and it invokes a system call that causes a "crash", should I ever expect to abort another, unrelated process as side effect or, worse, the entire operating system?  To do so is opening the architecture to denial of service attacks.  Wasmtime is the operating system to these Wasm processes and it should not allow for that either (within reason).

How users may link in the `wasmtime` crate via a Rust program with `panic=abort` shouldn't concern us, as they're opting into that behavior.  What should concern us is how we build the *wasmtime shared library* exposing the C API that will be dynamically linked to from many different languages.  It currently unwinds on panics and that allows us to properly translate panics into traps at the API boundary.

For correctness, that API must return a trap rather than unwind, so we're going to have to catch the panic *again* at the C API boundary so that functions like `wasm_func_call` can have the expected behavior.  Personally, I think it naturally follows that this would be desirable behavior of the Rust API too: when you invoke a Wasm function and it somehow triggers a panic, it is as if a Wasm function trapped and an `Err(Trap)` is returned.  Without this, we force users to use `catch_unwind` to implement the same robustness, which is not a good developer experience in my opinion.

If there's concern that we'll somehow lose debugability by having a trap rather than a panic for bugs in Wasmtime, I would argue that traps must be *at least* as easy to debug otherwise our the developer story is not a good one.

I still propose that resuming the panic and unwinding the Rust frames just to catch the panic again a few frames later simply to return a trap representation adds additional complexity, whereas storing a trap representation here and letting the trampoline trap would, at a minimum, remove the scariness of `longjmp`'ing through a Rust frame and the potential for future bugs as a result (your *there be dragons* comment notwithstanding).

I'm not married to my approach at all, but I really think we should give this a lot of consideration given all the different use cases for embedding Wasmtime.

Reply:
Alex and I discussed this in person.

The approach we'll take is that we'll do trap translation at the `wasm_func_call` boundary so that the C API will see a trap returned for panics.

For users of the Rust API, they can manually catch the panic if desired.  We may, in the future, offer a cleaner `Callable` API that handles a panic to trap translation in the future, but that's not necessary today.